src/Microsoft.SqlTools.ResourceProvider.DefaultImpl/AzureAuthenticationManager.cs

```diff
@@ -0,0 +1,314 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.SqlTools.ResourceProvider.Core;
+using Microsoft.SqlTools.ResourceProvider.Core.Authentication;
+using Microsoft.SqlTools.ResourceProvider.Core.Contracts;
+using Microsoft.SqlTools.ResourceProvider.Core.Extensibility;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ResourceProvider.DefaultImpl
+{
+
+    /// <summary>
+    /// Implementation for <see cref="IAzureAuthenticationManager" />. 
+    /// Provides functionality to authenticate to Azure and discover associated accounts and subscriptions
+    /// </summary>
+    [Exportable(
+        ServerTypes.SqlServer,
+        Categories.Azure,
+        typeof(IAzureAuthenticationManager),
+        "Microsoft.SqlTools.ResourceProvider.DefaultImpl.AzureAuthenticationManager",
+        1)
+    ]
+    class AzureAuthenticationManager : ExportableBase, IAzureAuthenticationManager
+    {
+        private Dictionary<string, AzureUserAccount> accountsMap;
+        private string currentAccountId = null;
+        private IEnumerable<IAzureUserAccountSubscriptionContext> _selectedSubscriptions = null;
```

Comment:
Can this be implemented as a Concurrent something to avoid the manual locking?

Reply:
I've ported this as-is from Leila's other codebase and am loathe to change right now. Let's look at this in the future, for now it's highly unlikely to actually lock.